src/effects/native/autopaneffect.cpp

```diff
@@ -0,0 +1,196 @@
+                              const unsigned int sampleRate,
+                              const EffectProcessor::EnableState enableState,
+                              const GroupFeatureState& groupFeatures) {
+    Q_UNUSED(group);
+    Q_UNUSED(groupFeatures);
+    Q_UNUSED(sampleRate);
+    
+    PanGroupState& gs = *pGroupState;
+    
+    if (enableState == EffectProcessor::DISABLED) {
+        return;
+    }
+    
+    CSAMPLE period = roundf(m_pPeriodParameter->value());
+    if (groupFeatures.has_beat_length) {
+        // 1/8, 1/4, 1/2, 1, 2, 4, 8, 16, 32, 64
+        double beats = pow(2, floor(period * 9 / 500) - 3);
+        period = groupFeatures.beat_length * beats;
+    } else {
+        period *= (float)numSamples / 2.0f;
+    }
+    
+    CSAMPLE stepFrac = m_pStrengthParameter->value();
+    CSAMPLE depth = m_pDepthParameter->value();
+    float rampingTreshold = m_pRampingParameter->value();
+    
+    if (gs.time > period || enableState == EffectProcessor::ENABLING) {
+        gs.time = 0;
+    }
+    
+    // Normally, the position goes from 0 to 1 linearly. Here we make steps at
+    // 0.25 and 0.75 to have the sound fully on the right or fully on the left.
+    // At the end, the "position" value can describe a sinusoid or a square
+    // curve depending of the size of those steps.
+    
+    // coef of the slope
+    // a = (y2 - y1) / (x2 - x1)
+    //       1  / ( 1 - 2 * stepfrac)
+    float a = stepFrac != 0.5f ? 1.0f / (1.0f - stepFrac * 2.0f) : 1.0f;
+    
+    // define the increasing of gain (the music sounds lower if only one 
+    // channel is enabled).
+    // TODO(jclaveau) : Challenge this value
+    float lawCoef = 1.0f / sqrtf(2.0f) * 2.0f;
+    
+    // size of a segment of slope (controled by the "strength" parameter)
+    float u = (0.5f - stepFrac) / 2.0f;
+    
+    gs.frac.setRamping(rampingTreshold);
+    gs.frac.ramped = false;     // just for debug
+    
+    
+    for (unsigned int i = 0; i + 1 < numSamples; i += 2) {
+        
+        CSAMPLE periodFraction = CSAMPLE(gs.time) / period;
+        
+        // current quarter in the trigonometric circle
+        float quarter = floorf(periodFraction * 4.0f);
+        
+        // part of the period fraction being a step (not in the slope)
+        CSAMPLE stepsFractionPart = floorf((quarter+1.0f)/2.0f) * stepFrac;
+        
+        // float inInterval = fmod( periodFraction, (period / 2.0) );
+        float inStepInterval = fmod(periodFraction, 0.5f);
+        
+        CSAMPLE position;
+        if (inStepInterval > u && inStepInterval < (u + stepFrac)) {
+            // at full left or full right
+            position = quarter < 2.0f ? 0.25f : 0.75f;
+        } else {
+            // in the slope (linear function)
+            position = (periodFraction - stepsFractionPart) * a;
+        }
+        
+        // transform the position into a sinusoid (but between 0 and 1)
+        gs.frac = (sin(M_PI * 2.0f * position) + 1.0f) / 2.0f;
```

Comment:
Mm, I do not understand how this will work in any case. Probably there are still click issues. 
If clicking occurs during parameter change, we have nice functions that fade from on value to an other used in the other effects.